File: facebook_client.py
import os
import logging
import httpx
from dotenv import load_dotenv

load_dotenv()

logger = logging.getLogger(__name__)

FB_VERIFY_TOKEN = os.getenv("FB_VERIFY_TOKEN", "fcommerce_ai_secret_token_123")
FB_PAGE_ACCESS_TOKEN = os.getenv("FB_PAGE_ACCESS_TOKEN", "")

class FacebookMessengerClient:
    def __init__(self, page_access_token: str = None):
        self.access_token = page_access_token or FB_PAGE_ACCESS_TOKEN
        self.api_url = "https://graph.facebook.com/v19.0/me/messages"

    def verify_webhook(self, mode: str, token: str, challenge: str) -> str | None:
        """Verifies the webhook token sent by Meta during endpoint registration."""
        if mode == "subscribe" and token == FB_VERIFY_TOKEN:
            logger.info("Facebook Webhook verified successfully!")
            return challenge
        logger.warning(f"Webhook verification failed. Given token: {token}")
        return None

    def extract_incoming_messages(self, payload: dict) -> list[dict]:
        """Extracts sender PSID, Page ID, and text content from incoming Meta webhook payload."""
        messages = []
        try:
            entry_list = payload.get("entry", [])
            for entry in entry_list:
                page_id = entry.get("id")
                messaging_events = entry.get("messaging", [])
                for event in messaging_events:
                    sender_id = event.get("sender", {}).get("id")
                    recipient_id = event.get("recipient", {}).get("id")
                    
                    # Handle normal text message
                    if "message" in event and "text" in event["message"]:
                        text = event["message"]["text"]
                        messages.append({
                            "psid": sender_id,
                            "page_id": page_id,
                            "recipient_id": recipient_id,
                            "text": text,
                            "type": "text"
                        })
                    # Handle postback button clicks
                    elif "postback" in event and "payload" in event["postback"]:
                        payload_text = event["postback"]["payload"]
                        messages.append({
                            "psid": sender_id,
                            "page_id": page_id,
                            "recipient_id": recipient_id,
                            "text": payload_text,
                            "type": "postback"
                        })
        except Exception as e:
            logger.error(f"Error parsing Facebook webhook payload: {e}")
        return messages

    async def send_text_message(self, psid: str, text: str) -> bool:
        """Sends a text reply to a customer via Facebook Graph API."""
        if not self.access_token or self.access_token == "mock_page_access_token":
            logger.info(f"[TEST MODE - Mock FB Send] To PSID: {psid} | Message: {text}")
            return True

        params = {"access_token": self.access_token}
        payload = {
            "recipient": {"id": psid},
            "messaging_type": "RESPONSE",
            "message": {"text": text}
        }

        async with httpx.AsyncClient() as client:
            try:
                response = await client.post(self.api_url, params=params, json=payload, timeout=10.0)
                if response.status_code == 200:
                    logger.info(f"Successfully sent FB message to PSID {psid}")
                    return True
                else:
                    logger.error(f"Failed to send FB message: {response.status_code} - {response.text}")
                    return False
            except Exception as e:
                logger.error(f"HTTP error sending FB message: {e}")
                return False

File: facebookClient.js
import dotenv from 'dotenv';
dotenv.config();

const FB_VERIFY_TOKEN = process.env.FB_VERIFY_TOKEN || "fcommerce_ai_secret_token_123";
const FB_PAGE_ACCESS_TOKEN = process.env.FB_PAGE_ACCESS_TOKEN || "";

export class FacebookMessengerClient {
  constructor(accessToken = null) {
    this.accessToken = accessToken || FB_PAGE_ACCESS_TOKEN;
    this.apiUrl = "https://graph.facebook.com/v19.0/me/messages";
  }

  async getPageDetails(pageId = null, pageAccessToken = null) {
    const token = pageAccessToken || this.accessToken || process.env.FB_PAGE_ACCESS_TOKEN;
    if (!token || token.includes("mock_page_access_token") || token.trim() === "") {
      return null;
    }

    const targetId = pageId || "me";
    try {
      const res = await fetch(`https://graph.facebook.com/v19.0/${targetId}?access_token=${token}&fields=name,id`);
      if (res.ok) {
        const data = await res.json();
        console.log(`✅ Fetched Meta Page Details for ${data.id}: "${data.name}"`);
        return data;
      } else {
        const errText = await res.text();
        console.warn(`Could not fetch Meta page details (${res.status}):`, errText);
      }
    } catch (e) {
      console.error("Error fetching Page details from Meta:", e);
    }
    return null;
  }

  verifyWebhook(mode, token, challenge) {
    if (mode === "subscribe" && token === FB_VERIFY_TOKEN) {
      console.log("✅ Facebook Webhook verified successfully!");
      return challenge;
    }
    console.warn(`⚠️ Webhook verification failed. Received token: ${token}`);
    return null;
  }

  extractIncomingMessages(payload) {
    const messages = [];
    try {
      const entryList = payload.entry || [];
      for (const entry of entryList) {
        const pageId = entry.id;
        const messagingEvents = entry.messaging || [];
        for (const event of messagingEvents) {
          const senderId = event.sender?.id;
          const recipientId = event.recipient?.id;

          if (event.message && event.message.text) {
            messages.push({
              psid: senderId,
              pageId,
              recipientId,
              text: event.message.text,
              type: 'text'
            });
          } else if (event.postback && event.postback.payload) {
            messages.push({
              psid: senderId,
              pageId,
              recipientId,
              text: event.postback.payload,
              type: 'postback'
            });
          }
        }
      }
    } catch (err) {
      console.error("Error parsing Facebook webhook payload:", err);
    }
    return messages;
  }

  async sendTextMessage(psid, text, overrideToken = null) {
    const token = overrideToken || this.accessToken || process.env.FB_PAGE_ACCESS_TOKEN;

    if (!token || token.includes("mock_page_access_token") || token.trim() === "") {
      console.warn(`⚠️ Cannot send live Meta message: Page Access Token is missing or set to mock token! (PSID: ${psid})`);
      console.log(`[SIMULATED AI RESPONSE TO PSID ${psid}]:\n${text}`);
      return false;
    }

    try {
      const response = await fetch(`${this.apiUrl}?access_token=${token}`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          recipient: { id: psid },
          messaging_type: "RESPONSE",
          message: { text }
        })
      });

      if (response.ok) {
        console.log(`✅ Successfully sent live FB Messenger reply to PSID ${psid}`);
        return true;
      } else {
        const errorText = await response.text();
        console.error(`❌ Meta Graph API Error (${response.status}):`, errorText);
        return false;
      }
    } catch (err) {
      console.error("HTTP error sending FB message:", err);
      return false;
    }
  }
}

export const fbClient = new FacebookMessengerClient();
